Fix behavior of select type custom profile fields when deleting and reordering options.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit adds code to delete `CustomProfileFieldValue` objects when deleting an option.
- Second commit adds migration to delete existing cases of `CustomProfileFieldValue` rows with invalid values.
- Third commit updates web-app code to keep value of the option same irrespective of reordering and deleting options.

Fixes: <!-- Issue link, or clear description.-->#22145

Discussion - https://chat.zulip.org/#narrow/stream/378-api-design/topic/custom.20profile.20fields.20option.20deletion

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented:
Added commit to show modal when deleting option. Sharing a GIF below -
![custom-field](https://user-images.githubusercontent.com/35494118/174797370-a0fc7303-7ad9-491c-a42a-9e6d18a78532.gif)


We currently do not show modal when deleting option while creating a new profile field or when removing and newly added option when editing. Is this fine or we want to do something different?

sahil839 commented on static/js/settings_profile_fields.js:
```diff
@@ -105,8 +106,23 @@ function update_choice_delete_btn($container, display_flag) {
+        existing_option_values.push(Number.parseInt($(this).attr("data-value"), 10));
+    });
+    existing_option_values.sort();
+    return existing_option_values[existing_option_values.length - 1] + 1;
```
One other thing we can do here, rather than using the value greater than the already used largest value, we can instead use first number not used already starting from "0".

sahil839 replied:
We cannot do this since we want to handle the case where an existing option is deleted, then a new option added takes value of that deleted option and then we again add that option as mentioned [here](https://github.com/zulip/zulip/pull/22267/commits/d0d62530251fbf8b10c6488fda27346d41eb4975#r915077392) meaning same values for two options.

sahil839 commented on static/js/settings_profile_fields.js:
```diff
@@ -231,7 +247,7 @@ export function parse_field_choices_from_field_data(field_data) {
         });
     }
-
+    choices.sort((a, b) => a.order - b.order);
```
We need this because currently the choices are returned from server in the ascending order of values/id of the option. We can skip this if we remove order field and instead pass a list as suggested in https://chat.zulip.org/#narrow/stream/378-api-design/topic/custom.20profile.20fields/near/1382989.

sahil839 commented on static/templates/confirm_dialog/confirm_delete_profile_field_option.hbs:
```diff
@@ -0,0 +1,3 @@
+{{#tr}}
+    This will delete the field for {{count}} users. Are you sure you want to delete this option?
```
Any suggestions to improve the text here or is this fine?

alya commented:
We should adjust the message to have the correct form when N = 1 ("1 user", not "1 users").

<img width="544" alt="Screen Shot 2022-06-29 at 11 15 54 AM" src="https://user-images.githubusercontent.com/2090066/176509318-7fc351ef-4c7a-4dd7-b891-cdcc8cc95082.png">

sahil839 commented:
Ok, working on it.

sahil839 commented on static/js/settings_profile_fields.js:
```diff
@@ -66,15 +66,27 @@ function delete_profile_field(e) {
             if (text) {
-                field_data[field_order - 1] = {text, order: field_order.toString()};
+                if (old_option_value_map.get(text) !== undefined) {
+                    $(this).attr("data-value", old_option_value_map.get(text));
```
This code is added to make sure that in a case when I delete one existing option, but add same option again before saving and then save the changes, then user who had already selected that value retain their selection.

timabbott replied:
OK, let's move this into a comment, since it's pretty subtle.

sahil839 commented:
Updated the PR.
- Changed the behavior to show modal after clicking on `Save changes` and also updated the text as asked above.
- Also added code to handle the case when deleting an existing option and adding it again as mentioned [here](https://github.com/zulip/zulip/pull/22267#discussion_r915077392).
- Also added commit to update the data used in our development and test database to be in the same format as sent by webapp.

sahil839 commented:
Updated the PR.
- Changed the behavior to show modal after clicking on `Save changes` and also updated the text as asked above.
- Also added code to handle the case when deleting an existing option and adding it again as mentioned [here](https://github.com/zulip/zulip/pull/22267#discussion_r915077392).
- Also added commit to update the data used in our development and test database to be in the same format as sent by webapp.

timabbott commented:
Merged, after renumbering the migration and adding the comment suggested above. Thanks @sahil839!

I wonder if there's an easy refactor that would make `populate_db` validate that custom profile field values being set match the format -- that `populate_db` bug was pretty nasty.
